### PR TITLE
Merge ML dev

### PR DESCRIPTION
**Pull request by kaminow**
Finally merge the ML infrastructure into `main`. Still in the process of updating the `ml-dev` branch to the new directory structure.

Some important things that have been added:
- Start using [Weights & Biases](https://wandb.ai/) for tracking model training and for hyperparameter sweeps
- Add support for multi-structure ("Grouped") datasets
- Organize and centralize functions to load data and build models
- Implement early stopping

### CODE REVIEW

**Review by hmacdope — APPROVED** [review]
Looks good to me @kaminow :)

**Comment by hmacdope** [comment]
I will merge this once packaging is sorted out for `mtenn`  and I can update the conda env accordingly.

**Review by kaminow — COMMENTED** [review]
two quick questions, but otherwise let's merge this!

File: devtools/conda-envs/asapdiscovery.yaml [hunk 1]
```diff
@@ -1,10 +1,9 @@
 name: asapdiscovery
 channels:
-  - pytorch
-  - pyg
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply devtools/conda-envs/asapdiscovery.yaml:4]: The packages are on `conda-forge`.  confusingly with slightly different names (eg `pytorch_geometric` on `conda-forge`  channel and `pyg` on `pyg` channel etc).

Using the conda-forge spec was the only way I could easily get everything cooperating as they are all cross tested there.

[the remark this answers is not in the document]
Comment (kaminow) [reply devtools/conda-envs/asapdiscovery.yaml:4]: sounds good!

**Pull request merged**